File: HearthVale/client/src/utils/mapBounds.ts
import type { MapDefinition } from '../types/world.js';
import { TILE_SIZE } from '../constants.js';

export interface MapBounds {
  minX: number;
  maxX: number;
  minY: number;
  maxY: number;
}

/** Map bounds from gridSize — world origin at center. */
export function getMapBounds(map: MapDefinition): MapBounds {
  const halfW = (map.gridSize.width * TILE_SIZE) / 2;
  const halfH = (map.gridSize.height * TILE_SIZE) / 2;

  let minX = -halfW;
  let maxX = halfW;
  let minY = -halfH;
  let maxY = halfH;

  const points = [
    map.playerSpawn,
    ...map.portals.map((p) => p.position),
    ...map.npcs.map((n) => n.position),
  ];

  for (const point of points) {
    minX = Math.min(minX, point.x - TILE_SIZE);
    maxX = Math.max(maxX, point.x + TILE_SIZE);
    minY = Math.min(minY, point.y - TILE_SIZE);
    maxY = Math.max(maxY, point.y + TILE_SIZE);
  }

  return { minX, maxX, minY, maxY };
}

export function clampToBounds(x: number, y: number, bounds: MapBounds): { x: number; y: number } {
  return {
    x: Math.min(Math.max(x, bounds.minX), bounds.maxX),
    y: Math.min(Math.max(y, bounds.minY), bounds.maxY),
  };
}

File: HearthVale/client-unity/Assets/Scripts/World/MapBounds.cs
// MapBounds.cs
// C# port of client-phaser-archive/src/utils/mapBounds.ts.
// World origin is centered at (0,0), same convention as the Phaser client, so
// existing map/portal/NPC coordinates in data/maps.json need no transform.

using System.Collections.Generic;
using HearthVale.Data;
using UnityEngine;

namespace HearthVale.World
{
    public struct MapBoundsRect
    {
        public float minX, maxX, minY, maxY;
    }

    public static class MapBounds
    {
        /// <summary>Tile size in world units. Matches TILE_SIZE (32px) in the
        /// archived client's constants.ts, reinterpreted as Unity world units
        /// via WorldConstants.TileSize (see WorldConstants.cs).</summary>
        public static MapBoundsRect GetMapBounds(MapDefinition map)
        {
            float halfW = (map.gridSize.width * WorldConstants.TileSize) / 2f;
            float halfH = (map.gridSize.height * WorldConstants.TileSize) / 2f;

            float minX = -halfW, maxX = halfW, minY = -halfH, maxY = halfH;

            var points = new List<Vec2> { map.playerSpawn };
            foreach (var p in map.portals) points.Add(p.position);
            foreach (var n in map.npcs) points.Add(n.position);

            foreach (var point in points)
            {
                minX = Mathf.Min(minX, point.x - WorldConstants.TileSize);
                maxX = Mathf.Max(maxX, point.x + WorldConstants.TileSize);
                minY = Mathf.Min(minY, point.y - WorldConstants.TileSize);
                maxY = Mathf.Max(maxY, point.y + WorldConstants.TileSize);
            }

            return new MapBoundsRect { minX = minX, maxX = maxX, minY = minY, maxY = maxY };
        }

        public static Vector2 ClampToBounds(float x, float y, MapBoundsRect bounds)
        {
            return new Vector2(
                Mathf.Clamp(x, bounds.minX, bounds.maxX),
                Mathf.Clamp(y, bounds.minY, bounds.maxY));
        }
    }
}
